### PR TITLE
missing sensor api's

### PR DESCRIPTION
**Pull request by pvschakradhar**
### Description of Change ###
Added Missing Sensor API's and Sensor Classes

### API Changes ###
 - ACR: http://suprem.sec.samsung.net/jira/browse/TCSACR-218

### CODE REVIEW

File: src/Tizen.Sensor/Tizen.Sensor/EventArgs/ProviderIntervalChangedEventArgs.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        }
+
+        /// <summary>
+        /// Interval
+        /// </summary>
+        public uint IntervalMs { get; private set; }
+        /// <summary>
+        /// Indicates userdata sent by provider Interval changed callback
+        /// </summary>
+        public IntPtr UserData { get; private set; }
```
Comment (m-macias) [inline src/Tizen.Sensor/Tizen.Sensor/EventArgs/ProviderIntervalChangedEventArgs.cs]: User shouldn't have access to IntPtr. Please wrap it up to some SafeHandle or redesign this class.

File: src/Tizen.Sensor/Tizen.Sensor/Sensor.cs [hunk 2]
```diff
@@ -545,5 +604,407 @@ private void DestroyHandles()
...
+        public void RecorderCreateOption() {
+            int error = Interop.SensorRecoder.RecorderCreateOption(out recorderOption);
+            if (error != (int)SensorError.None)
+            {
+                Log.Error(Globals.LogTag, "Error creating the sensor recorder option");
+                throw SensorErrorFactory.CheckAndThrowException(error, "Sensor.RecorderCreateOption Failed");
+            }
+        }
+
+        public void RecorderCreateQuery()
```
Comment (m-macias) [inline src/Tizen.Sensor/Tizen.Sensor/Sensor.cs]: Public methods should have a documentation.

File: src/Tizen.Sensor/Tizen.Sensor/EventArgs/CustomSensorDataUpdatedEventArgs.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    public class CustomSensorDataUpdatedEventArgs : EventArgs
+    {
+        internal CustomSensorDataUpdatedEventArgs(Interop.SensorEventStruct sensorData)
+        {
+            try
+            {
+                Values = new float[sensorData.value_count];
+                Array.Copy(sensorData.values, Values, sensorData.value_count);
+            }
+            catch (OutOfMemoryException e)
```
Comment (myroot) [inline src/Tizen.Sensor/Tizen.Sensor/EventArgs/CustomSensorDataUpdatedEventArgs.cs]: Do not catch OutOfMemoryException.
You need notify error situation to the app

File: src/Tizen.Sensor/Tizen.Sensor/EventArgs/HeartRateMonitorLedGreenDataUpdatedEventArgs.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+using System;
+
+
+namespace Tizen.Sensor
+{
+    /// <summary>
+    /// The HeartRateMonitorLedGreen changed event arguments class is used for storing the data returned by a heart rate monitor Led Green.
+    /// </summary>
+    /// <since_tizen> 6 </since_tizen>
+    public class HeartRateMonitorLedGreenDataUpdatedEventArgs : EventArgs
```
Comment (myroot) [inline src/Tizen.Sensor/Tizen.Sensor/EventArgs/HeartRateMonitorLedGreenDataUpdatedEventArgs.cs]: Do you need distinguish a type from HeartRateMonitorDataUpdatedEventArgs?
Why?

File: src/Tizen.Sensor/Tizen.Sensor/EventArgs/ProviderIntervalChangedEventArgs.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        /// </summary>
+        internal ProviderIntervalChangedEventArgs(Int64 _userData, uint intervalMs) {
+            IntervalMs = intervalMs;
+            UserData = _userData;
+        }
+
+        /// <summary>
+        /// Interval
+        /// </summary>
+        public uint IntervalMs { get; private set; }
```
Comment (myroot) [inline src/Tizen.Sensor/Tizen.Sensor/EventArgs/ProviderIntervalChangedEventArgs.cs]: Add more description about time unit
```suggestion
        public uint Interval { get; private set; }
```

File: src/Tizen.Sensor/Tizen.Sensor/EventArgs/ProviderIntervalChangedEventArgs.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        /// <summary>
+        /// Indicates userdata sent by provider Interval changed callback
+        /// </summary>
+        public Int64 UserData { get; private set; }
```
Comment (myroot) [inline src/Tizen.Sensor/Tizen.Sensor/EventArgs/ProviderIntervalChangedEventArgs.cs]: No need user data

File: src/Tizen.Sensor/Tizen.Sensor/Sensor.cs [hunk 7]
```diff
@@ -545,5 +604,600 @@ private void DestroyHandles()
...
+            int error = Interop.SensorProvider.ProviderSetStopCb(_sensorProviderHandle, _sensorProviderStopCallBack, userData);
+            if (error != (int)SensorError.None)
+            {
+                Log.Error(Globals.LogTag, "Error in setting stop callback of sensor provider");
+                throw SensorErrorFactory.CheckAndThrowException(error, "Sensor.ProviderSetStopCb Failed");
+            }
+
+        }
+
+        private void ProviderSetIntervalChangedCb(IntPtr userData)
```
Comment (myroot) [inline src/Tizen.Sensor/Tizen.Sensor/Sensor.cs]: Who is use it?
I can't found. 
And it is private

File: src/Tizen.Sensor/Tizen.Sensor/Sensor.cs [hunk 8]
```diff
@@ -545,5 +604,600 @@ private void DestroyHandles()
...
+        }
+
+        /// <summary>
+        /// Adds a callback function to be invoked when a new sensor is added.
+        /// </summary>
+        /// <param name="userData">
+        /// The user data to be passed to the callback function.
+        /// </param>
+        /// <since_tizen> 6 </since_tizen>
+        public void AddSensorAddedCB(IntPtr userData)
```
Comment (myroot) [inline src/Tizen.Sensor/Tizen.Sensor/Sensor.cs]: USE event, DO NOT USE callback

**Comment by myroot** [comment]
If do not give any feedback, I will close it on today

File: src/Tizen.Sensor/Tizen.Sensor/EventArgs/HeartRateMonitorLedGreenDataUpdatedEventArgs.cs (hunk 4 above)
Comment (pvschakradhar) [reply src/Tizen.Sensor/Tizen.Sensor/EventArgs/HeartRateMonitorLedGreenDataUpdatedEventArgs.cs]: Removed  HeartRateMonitorLedGreenDataUpdatedEventArgs

File: src/Tizen.Sensor/Tizen.Sensor/EventArgs/ProviderIntervalChangedEventArgs.cs (hunk 5 above)
Comment (pvschakradhar) [reply src/Tizen.Sensor/Tizen.Sensor/EventArgs/ProviderIntervalChangedEventArgs.cs]: updated to Interval.

File: src/Tizen.Sensor/Tizen.Sensor/EventArgs/ProviderIntervalChangedEventArgs.cs (hunk 6 above)
Comment (pvschakradhar) [reply src/Tizen.Sensor/Tizen.Sensor/EventArgs/ProviderIntervalChangedEventArgs.cs]: removed UserData

File: src/Tizen.Sensor/Tizen.Sensor/Sensor.cs (hunk 7 above)
Comment (pvschakradhar) [reply src/Tizen.Sensor/Tizen.Sensor/Sensor.cs]: ProviderSetIntervalChangedCb access level changed to public

File: src/Tizen.Sensor/Tizen.Sensor/Sensor.cs (hunk 8 above)
Comment (pvschakradhar) [reply src/Tizen.Sensor/Tizen.Sensor/Sensor.cs]: API name is changed to SetSensorAddedCB. This is API for setting callback .

File: src/Tizen.Sensor/Tizen.Sensor/EventArgs/CustomSensorDataUpdatedEventArgs.cs (hunk 3 above)
Comment (pvschakradhar) [reply src/Tizen.Sensor/Tizen.Sensor/EventArgs/CustomSensorDataUpdatedEventArgs.cs]: Try catch is replaced with error notification

File: src/Tizen.Sensor/Tizen.Sensor/EventArgs/ProviderStartEventArgs.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Tizen.Sensor
+{
+    /// <summary>
+    /// Indicates Provider Start callback.
+    /// </summary>
+    public class ProviderStartEventArgs : EventArgs
```
Comment (myroot) [inline src/Tizen.Sensor/Tizen.Sensor/EventArgs/ProviderStartEventArgs.cs]: Do not make EventArgs that do not provide anything

File: src/Tizen.Sensor/Tizen.Sensor/EventArgs/ProviderStopEventArgs.cs [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Tizen.Sensor
+{
+    /// <summary>
+    /// Indicates Provider stop callback.
+    /// </summary>
+    public class ProviderStopEventArgs : EventArgs
```
Comment (myroot) [inline src/Tizen.Sensor/Tizen.Sensor/EventArgs/ProviderStopEventArgs.cs]: Do not make EventArgs that do not provide anything

File: src/Tizen.Sensor/Tizen.Sensor/EventArgs/RecorderDataEventArgs.cs [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+        /// <summary>
+        /// Gets Remains.
+        /// </summary>
+        public int Remains { get; private set; }
+
+        /// <summary>
+        /// Gets Error.
+        /// </summary>
+        public Enum Error { get; private set; }
```
Comment (myroot) [inline src/Tizen.Sensor/Tizen.Sensor/EventArgs/RecorderDataEventArgs.cs]: Do not use Error code

File: src/Tizen.Sensor/Tizen.Sensor/EventArgs/SensorAddedEventArgs.cs [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Tizen.Sensor
+{
+    /// <summary>
+    /// Indicates Sensor Added callback.
+    /// </summary>
+    public class SensorAddedEventArgs : EventArgs
```
Comment (myroot) [inline src/Tizen.Sensor/Tizen.Sensor/EventArgs/SensorAddedEventArgs.cs]: Do not make EventArgs that do not provide anything

File: src/Tizen.Sensor/Tizen.Sensor/Sensor.cs [hunk 13]
```diff
@@ -545,5 +652,606 @@ private void DestroyHandles()
...
+        /// Checks whether it is supported to record a given sensor type.
+        /// </summary>
+        /// <param name="type">
+        /// A sensor type to check.
+        /// </param>
+        /// <param name="isSupported">
+        /// If supported.
+        /// </param>
+        /// <since_tizen> 6 </since_tizen>
+        public void RecorderIsSupported(Enum type, out IntPtr isSupported)
```
Comment (myroot) [inline src/Tizen.Sensor/Tizen.Sensor/Sensor.cs]: Do not use IntPtr in public API

Why do you use `System.Enum` class?

File: src/Tizen.Sensor/Tizen.Sensor/Sensor.cs [hunk 14]
```diff
@@ -24,12 +25,60 @@ internal static class Globals
         internal const string LogTag = "Tizen.Sensor";
     }
 
+    /// <summary>
+    /// Data from sensor
+    /// </summary>
+    [StructLayout(LayoutKind.Sequential, Pack = 0)]
+    public struct SensorEventStruct
```
Comment (myroot) [inline src/Tizen.Sensor/Tizen.Sensor/Sensor.cs]: Do not directly expose native data struct

File: src/Tizen.Sensor/Tizen.Sensor/Sensor.cs [hunk 15]
```diff
@@ -545,5 +652,606 @@ private void DestroyHandles()
...
+        /// Gets the handle for the default sensor of a given sensor URI.
+        /// </summary>
+        /// <param name="uri">
+        /// A sensor or a sensor type URI to get the handle of its default sensor.
+        /// </param>
+        /// <param name="sensorHandle">
+        /// The default sensor handle.
+        /// </param>
+        /// <since_tizen> 6 </since_tizen>
+        public void GetDefaultSensorByUri(string uri, out IntPtr sensorHandle)
```
Comment (myroot) [inline src/Tizen.Sensor/Tizen.Sensor/Sensor.cs]: Do not use IntPtr and out param in public API

File: src/Tizen.Sensor/Tizen.Sensor/Sensor.cs [hunk 16]
```diff
@@ -545,5 +652,606 @@ private void DestroyHandles()
...
+                throw SensorErrorFactory.CheckAndThrowException(error, "Sensor.ProviderSetResolution Failed");
+            }
+
+        }
+
+        /// <summary>
+        /// Registers the callback function to be invoked when a listener starts the sensor provider.
+        /// </summary>
+        /// <since_tizen> 6 </since_tizen>
+        public void ProviderSetStartCb()
```
Comment (myroot) [inline src/Tizen.Sensor/Tizen.Sensor/Sensor.cs]: What is it?
Do not USE callback

**Comment by myroot** [comment]
Please, study about .net framework design guide https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/
And Please refer others TizenFX frameworks APIs, ex) Applications, Multimedia, 
Do not directly mapping API from C API

File: src/Tizen.Sensor/Tizen.Sensor/EventArgs/ProviderStartEventArgs.cs (hunk 9 above)
Comment (pvschakradhar) [reply src/Tizen.Sensor/Tizen.Sensor/EventArgs/ProviderStartEventArgs.cs]: Removed the empty EventArgs.

File: src/Tizen.Sensor/Tizen.Sensor/EventArgs/ProviderStopEventArgs.cs (hunk 10 above)
Comment (pvschakradhar) [reply src/Tizen.Sensor/Tizen.Sensor/EventArgs/ProviderStopEventArgs.cs]: Removed the empty EventArgs.

File: src/Tizen.Sensor/Tizen.Sensor/EventArgs/RecorderDataEventArgs.cs (hunk 11 above)
Comment (pvschakradhar) [reply src/Tizen.Sensor/Tizen.Sensor/EventArgs/RecorderDataEventArgs.cs]: Removed the additional variables in EventArgs.

File: src/Tizen.Sensor/Tizen.Sensor/EventArgs/SensorAddedEventArgs.cs (hunk 12 above)
Comment (pvschakradhar) [reply src/Tizen.Sensor/Tizen.Sensor/EventArgs/SensorAddedEventArgs.cs]: Removed the empty EventArgs.

File: src/Tizen.Sensor/Tizen.Sensor/Sensor.cs (hunk 13 above)
Comment (pvschakradhar) [reply src/Tizen.Sensor/Tizen.Sensor/Sensor.cs]: Inptr is removed from all public API's and Enum is replaced with int

File: src/Tizen.Sensor/Tizen.Sensor/Sensor.cs (hunk 15 above)
Comment (pvschakradhar) [reply src/Tizen.Sensor/Tizen.Sensor/Sensor.cs]: Inptr is removed from all public API's

File: src/Tizen.Sensor/Tizen.Sensor/Sensor.cs (hunk 16 above)
Comment (pvschakradhar) [reply src/Tizen.Sensor/Tizen.Sensor/Sensor.cs]: This API registers the callback function to be invoked when a listener starts the sensor provider.

File: src/Tizen.Sensor/Tizen.Sensor/Sensor.cs (hunk 14 above)
Comment (pvschakradhar) [reply src/Tizen.Sensor/Tizen.Sensor/Sensor.cs]: There is a public API "ProviderPublish" which has this "SensorEventStruct" as  parameter. So we created a public struct.

**Comment by myroot** [comment]
Your code is not meet a policy and principle of .net frameworks design guidelines and I already mentions about consistency with others API.
It was not matured to proceed a review process

File: src/Tizen.Sensor/Tizen.Sensor/Sensor.cs [hunk 17]
```diff
@@ -545,5 +638,623 @@ private void DestroyHandles()
...
+                throw SensorErrorFactory.CheckAndThrowException(error, "Sensor.ProviderPublish Failed");
+            }
+
+        }
+
+        /// <summary>
+        /// Adds a callback function to be invoked when a new sensor is added.
+        /// </summary>
+        /// <since_tizen> 6 </since_tizen>
+        public void SetSensorAddedCB()
```
Comment (myroot) [inline src/Tizen.Sensor/Tizen.Sensor/Sensor.cs]: 1. Do not use callback, use event
2. API name is `Set` but can't pass a value to set

File: src/Tizen.Sensor/Tizen.Sensor/Sensor.cs [hunk 18]
```diff
@@ -545,5 +638,623 @@ private void DestroyHandles()
...
+        {
+            _sensorAddedCallBack = (String uri, Int64 _userData) =>
+            {
```
Comment (myroot) [inline src/Tizen.Sensor/Tizen.Sensor/Sensor.cs]: implementation is empty

File: src/Tizen.Sensor/Tizen.Sensor/Sensor.cs [hunk 19]
```diff
@@ -545,5 +638,623 @@ private void DestroyHandles()
...
+        }
+
+        /// <summary>
+        ///  This function creates a sensor provider handle with a given URI.
+        /// </summary>
+        /// <param name="uri">
+        /// The URI of sensor to be created.
+        /// </param>
+        /// <since_tizen> 6 </since_tizen>
+        public void CreateProvider(string uri)
```
Comment (myroot) [inline src/Tizen.Sensor/Tizen.Sensor/Sensor.cs]: Expect something a created object, but return type is void. 
What is the means of Create?

**Comment by myroot** [comment]
Too many comments make slows down this page, so I close it.
If you have any update please make a new PR.

**Pull request closed without merging**